### PR TITLE
docs: add description of config versions

### LINKED ISSUE #20016: Document differences in config file schema versions
## Applies To

https://goteleport.com/docs/reference/config/

## Details

Currently, there are three schema versions that Teleport can use. These differences are not captured in the reference, which can exacerbate confusion around differences in Teleport behavior when changing the schema version.

This can be helped if the reference doc included the following:

* A note about the existence of different schema versions and why Teleport uses a versioned schema in the first place.
* A list of any schema-specific behaviors (v2 and v3 are multiplexing-on by default, for example)
* Any schema-exclusive options (auth_servers vs auth_server/proxy_server, for example)

Some information specific to the concept as a whole could get its own small note or section. Any other schema-specific options could appear in line with the reference config file doc. Some options already have some mentions of schema versions, but not all of them.

## Related Issues

#13324

### PR DESCRIPTION
**Pull request by zmb3**
This was originally added as part of #15761, but removed during the review cycle.

Closes #20016
Closes #23800

### CODE REVIEW

**Comment by zmb3** [comment]
@r0mant @smallinsky I would appreciate your review on the `v2` section. Hopefully I got it right :-)

File: docs/pages/reference/config.mdx [hunk 1]
```diff
@@ -168,3 +168,51 @@ These settings apply to the Windows Desktop Service:
...
+### Config `v1`
+
+`v1` is the original version of Teleport's file configuration. It is still supported
+today, but most new users should start with the latest configuration version.
+
+### Config `v2`
+
+Configuration version `v2` was introduced in Teleport 8 as part of Teleport's
+TLS routing feature. With TLS routing, Teleport's proxy listens on a single port
+and uses ALPN and SNI to route incoming traffic to the correct listener rather
```
Comment (smallinsky) [inline docs/pages/reference/config.mdx:189]: nit:
```suggestion
and uses ALPN and SNI to route incoming traffic to the correct Teleport service rather
```
Probably for most people listener=port so maybe we can use use `service` word instead.

**Review by smallinsky — APPROVED** [review]

**Review by ptgott — DISMISSED** [review]
I'm not sure about this approach. As a user visiting the configuration reference, I would be interested in specific fields and allowed values, which we don't get in the descriptions of the different versions. If I were interested in reverting to v1 of the config, for example, I'm not sure which fields I would need to change from v3. I suppose I could attempt to figure it out from the descriptions, but not with a great deal of confidence given how particular YAML is.

One thing we could do is turn `config.mdx` into a subsection of the "Reference" section, then include a version of this page in that subsection for each version of the file config.

If we expect most users to use v3 except for a certain subset with specific use cases, I would spell out that guidance explicitly instead of describing each version at a high conceptual level.

What do you think?

**Comment by zmb3** [comment]
I don't particularly care where we pit this, but you talked me out of it last time and there is still customer confusion so the only thing I feel strongly about is that this content makes it into the docs. 

We don't need to optimize for reverting to v1. Nobody should ever need to do that. The goal here is just to help answer the question of "why do I sometimes see different versions in the config file?"

**Comment by ptgott** [comment]
I would recommend phrasing this section as advice to users of specific config versions in the spirit of our [TLS Routing Migration](https://goteleport.com/docs/management/operations/tls-routing/) guide. Currently, it comes across as a high-level guide to each version. I'll make time to suggest a change.

**Comment by ptgott** [comment]
@zmb3 I've suggested edits in this PR: https://github.com/gravitational/teleport/pull/23800

The main change is organizing this section according to feature, rather than version number, so we can emphasize using the latest config version unless a user doesn't want to opt in to a specific feature.

**Comment by ptgott** [comment]
@zmb3 Thanks for the feedback on my PR. If the goal is to answer the question of what the difference is between `v2` and `v3`, would it make sense to address this question explicitly instead of adding a section explaining each configuration version, including `v1`? 

For me, it's a bit jarring to have a massive, detailed reference for `v3`, but only a few sentences about `v2` and `v1`. I see this leading to more questions than answers. Would it make sense to include a separate docs page to show `v1` and `v2` config users how to upgrade to `v3` instead?

In any case, I think we should get feedback from @alexfornuto or @xinding33 about this, since the config reference is such an important part of the docs. I'll remove my "Changes Requested" review so I'm not holding this up.

**Comment by zmb3** [comment]
@ptgott a few thoughts

> For me, it's a bit jarring to have a massive, detailed reference for v3, but only a few sentences about v2 and v1. 

This is somewhat intentional. All of our documentation, and even the code that generates configuration should always generate and refer to the latest version. Most new users don't care about v2 or v1, so there's less information on those.

If (when?) we add a v4 config, it wouldn't be unreasonable to remove most of the detail in the v3 section and replace it with a much shorter description so that the latest version gets most of the attention.

> would it make sense to address this question explicitly instead of adding a section explaining each configuration version, including v1?

That's the challenge. Config versions are not entirely new schemas - they're like 99% the same, with subtle differences. To understand the subtle differences, you kind of need to understand the behavior under the previous config version. That's why #20016 is requesting that we document the differences between versions.

In any case, this may not be perfect, but it's better than not explaining the config versions at all (which is the state we're in now). I'm always biased towards progress, and this feels an improvement over the current state to me.

File: docs/pages/reference/config.mdx (hunk 1 above)
Comment (alexfornuto) [reply docs/pages/reference/config.mdx:189]: It seems to me like listener or port would be correct. For example: when running a single Teleport service and connected to a database or openssh instance, the routing goes to a correct listener or port, not a correct Teleport service.

File: docs/pages/reference/config.mdx [hunk 2]
```diff
@@ -168,3 +168,51 @@ These settings apply to the Windows Desktop Service:
...
+
+For example, this excerpt from a `v2` config can be converted to `v3` with the
+following changes.
+
+```diff
+-version: v2
++version: v3
+teleport:
+-  auth_servers: [ teleport.example.com:443 ]
++  proxy_server: teleport.example.com:443
```
Comment (alexfornuto) [inline docs/pages/reference/config.mdx:217]: Could we make this example more descriptive where `auth_servers`, as an array, has two entries like `proxy.example.com` and `auth.example.com`, and `proxy_server` only has `proxy.example.com`?

Comment (zmb3) [reply docs/pages/reference/config.mdx:217]: We could, but specifying more than one (despite it being a list) has basically never worked correctly, so I'm a little hesitant to put an example in the docs for fear of someone copy pasting it. Thoughts?

**Comment by xinding33** [comment]
Hey folks, just chiming in here. I think we should merge this. It's not perfect but I view this change as a low risk and net positive as it adds more clarity than confusion.

**Review by xinding33 — APPROVED** [review]

File: docs/pages/reference/config.mdx (hunk 2 above)
Comment (alexfornuto) [reply docs/pages/reference/config.mdx:217]: Ah, that makes sense.

**Review by alexfornuto — APPROVED** [review]
I like it. It might be worth putting this above the "Reference configuration" header, since there's so much to scroll through in that section, but that's up to you.

**Pull request merged**